### ISSUE TITLE
Can't override headers per TransportOperation

### ISSUE THREAD

**Issue by SzymonPobiega** [Particular/NServiceBus#4502]
When creating multiple operations out of a single context, `RoutingToDispatchConnector` does not copy headers but passes the same instance to all operations. Because of that, if a behavior tries to modify headers in the dispatch stage, it modifies them for all the operations.

### Who's affected

Users who use transports with message-driven Pub/Sub and would like to attach destination-specific information in the headers (e.g. sequence number for a given destination)

### Workarounds

Use destination in the key for the destinations-specific header. The downside is that all the receivers get all information (more data on the wire)

### Solution

Copy the headers for each transport operation. Ideally we should copy-on-write them because in most cases they are not modified.

**Comment by SzymonPobiega** [comment]
@Particular/nservicebus-maintainers I am ready to help with solving this when if/when you prioritize it.

**Comment by timbussmann** [comment]
👍 to ensure we're not leaking header references between operations.
This seems to be a small fix we could deliver with 6.2.

**Comment by bording** [comment]
@SzymonPobiega Whats the actual use case here for wanting to modify the headers at this point? Why is this something we'd want to support? It feels strange to me to be wanting to modify the headers on the message after the logical message has been created.

And just to clarify, the only time when `RoutingToDispatchConnector` would be creating multiple operations is for publishes with message-driven pub/sub event, right?

If we did do this, would it matter that it would only be for message-driven pub/sub since native publishing would not have this option?

**Comment by SzymonPobiega** [comment]
@bording yet, the usage is related to message-driven pub/sub and the actual case is when you want to somehow mark the message flow in each particular queue for diagnostic/metric/statistic purposes. 

As there is a work-around I am OK with rejecting this issue but then I expect that in the next major the headers will be stored in a read-only manner after they are considered read-only.

**Comment by danielmarbach** [comment]
Btw. the same applies for the body. We create an outgoing message containing a reference to the body and the headers dictionary. The fix is unfortunately not so small because:

- Simple approach is to copy the dictionary then you play the price even though you are not actually modifying anything
- Copy on write is not so simple because outgoing message exposed `Dictionary<string, string>` and dictionary has no virtual overloads we could hook into to detect modifications.

**Comment by SzymonPobiega** [comment]
@danielmarbach switching to read-only dictionary could be an option. If transports want to add arbitrary header information, they would have to do it on the native level, when translating the headers dictionary into whatever the native form for headers is.

**Comment by danielmarbach** [comment]
@SzymonPobiega yeah but doing that in a minor means we have to support the old and the new way on the outgoing message and there is a perf penalty as well associated with the readonly dictionary.

**Comment by SzymonPobiega** [comment]
👍 for doing this in next major and marking this issue as breaking change.
